src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,144 @@
+    {
+        public CSharpUsePatternCombinatorsDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UsePatternCombinatorsDiagnosticId,
+                CSharpCodeStyleOptions.PreferPatternMatching,
+                LanguageNames.CSharp,
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.SwitchExpressionArm,
+                SyntaxKind.ConditionalExpression,
+                SyntaxKind.ForStatement,
+                SyntaxKind.EqualsValueClause,
+                SyntaxKind.IfStatement,
+                SyntaxKind.WhenClause,
+                SyntaxKind.WhileStatement,
+                SyntaxKind.DoStatement,
+                SyntaxKind.ReturnStatement,
+                SyntaxKind.YieldReturnStatement,
+                SyntaxKind.ArrowExpressionClause,
+                SyntaxKind.SimpleAssignmentExpression,
+                SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.ParenthesizedLambdaExpression,
+                SyntaxKind.Argument);
+
+        private static ExpressionSyntax? GetExpression(SyntaxNode node)
+            => node switch
+            {
+                SwitchExpressionArmSyntax n => n.Expression,
+                ConditionalExpressionSyntax n => n.Condition,
+                ForStatementSyntax n => n.Condition,
+                EqualsValueClauseSyntax n => n.Value,
+                IfStatementSyntax n => n.Condition,
+                WhenClauseSyntax n => n.Condition,
+                WhileStatementSyntax n => n.Condition,
+                DoStatementSyntax n => n.Condition,
+                ReturnStatementSyntax n => n.Expression,
+                YieldStatementSyntax n => n.Expression,
+                ArrowExpressionClauseSyntax n => n.Expression,
+                AssignmentExpressionSyntax n => n.Right,
+                LambdaExpressionSyntax n => n.ExpressionBody,
+                ArgumentSyntax n when n.GetRefKind() == RefKind.None => n.Expression,
+                _ => null,
+            };
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var parentNode = context.Node;
+            var syntaxTree = parentNode.SyntaxTree;
+            var cancellationToken = context.CancellationToken;
+
+            if (!((CSharpParseOptions)syntaxTree.Options).LanguageVersion.IsCSharp9OrAbove())
+                return;
+
+            var styleOption = context.Options.GetOption(CSharpCodeStyleOptions.PreferPatternMatching, syntaxTree, cancellationToken);
```

Comment:
i have a feeling we may need to break down this option into a few more.  i.e. PreferPatternMatching for not expression, for relational expressions, for binary expressions.

Reply:
>PreferPatternMatching for not expression, for relational expressions, for binary expressions

to be honest, I don't quite get why would you do that.

Reply:
because i think this is overbraod.  i.e. i can see people not wanting this for such idiomatic constructs as `x == 1 || x == 2`.

Reply:
>`x == 1 || x == 2`

It is idiomatic C# 1.0. coudn't we say `x is 1 or 2` would be idiomatic *now*?

btw my actual question was that why would you want separate options? if you prefer the old way you can just turn this off. I can't see why you want one of those and not the others. That's why the feature says 'Use pattern matching'. it's not clear to me why I care how my expression is turned to patterns. Either I want it or not.